### ISSUE TITLE
unable to bind kotlin.collections.List<Enum>

### ISSUE THREAD

**Issue by psartini** [jdbi/jdbi#1482]
Binding kotlin lists of enums do not work as expected.
Falling back to java.util.List works.

This is true for `List<Enum>`
`List<UUID>` works as expected.

please see https://github.com/psartini/jdbi-testcase-1482 for a test case.

Array is registered at
https://github.com/psartini/jdbi-testcase-1482/blob/112830ca7e730e07b74a9f3acde0a28ed9e594f1/src/test/kotlin/tools/JdbiProvider.kt#L31

(also see #1477 - this was the original issue)

**Comment by qualidafial** [comment]
Can you provide a list of Kotlin collection types that should be supported?

**Comment by psartini** [comment]
@qualidafial 
You mentioned in the other ticket it could be a missing CollectorFactory.
But List<UUID> is working without problems.

**Comment by qualidafial** [comment]
So I misspoke about the CollectorFactory being a solution to your particular case. Collectors only work for mapping, not binding.

The built in SQL array binding only supports Java arrays and Iterable types--surprisingly, Kotlin collections don't implement this interface in the JVM.

So adding support for binding Kotlin collections as SQL arrays would be a little more work.

**Comment by psartini** [comment]
Sorry, but I still don't get why it works with kotlin.collections.List<UUID> , 
but not with kotlin.collections.List<EnumType>

**Comment by FredDeschenes** [comment]
I've looked into this a bit and while I still have no idea what's happening, I can at least give you a workaround : it seems to work fine when using `Array` instead of `List` (in both the DAO and mapped return class):

```
data class DemoTableEnumWithKotlinArray(val demoId: UUID, val enumlist: Array<MyEnum>)

@SqlUpdate("INSERT INTO demotableenum (enumlist) VALUES (:enumList)")
@GetGeneratedKeys
fun insertKotlinEnumArray(enumList: Array<MyEnum>) : DemoTableEnumWithKotlinArray
```

**Comment by FredDeschenes** [comment]
Another thing I noticed that my queue somebody in to what's happening : looking at the DAO's .class file, the ones declared with Kotlin's `List` end up having their parameter declared as a `java.util.List<? extends MyEnum>`, while the ones declared with Java's `List` are as expected (`java.util.List<MyEnum>`).

The Kotlin one also never ends up in `SqlArrayArgumentFactory` (using 3.7.0-SNAPSHOT, I noticed this changed quite a bit since 3.6) for some reason, but I haven't found why yet.

**Comment by qualidafial** [comment]
It could be that the `out T` wildcard doesn't affect UUID because it is final, whereas enums are non-final to allow for anonymous inner classes per enum constant.

**Comment by qualidafial** [comment]
What happens if you change `List<MyEnum>` to `MutableList<MyEnum>`?

**Comment by psartini** [comment]
Yep - MutableList<MyEnum> works as well. Will add it to the test case.

**Comment by FredDeschenes** [comment]
By the way I've just validated that the same thing happens in Java : declaring a `List<? extends MyEnum>` as an argument will give the same result so it's not Kotlin-specific (although Kotlin adds the covariance stuff automatically).

It also seems that retrieving arrays directly via SQLObject doesn't work (unless I'm missing something).

**Comment by stevenschlansker** [comment]
This might be running into one of the known shortcomings of jdbi's type handling -- we currently key many things off of their raw `Class<?>` so we can do efficient Map lookups, but that doesn't handle wildcard types `? extends MyEnum` necessarily.

https://github.com/jdbi/jdbi/issues/1029

**Comment by stevenschlansker** [comment]
Also, I'm surprised it generates a `? extends MyEnum`, as `MyEnum` is guaranteed to be `final` as far as I understand it:

```
Compiled from "CompanyStatus.java"
public final class com.mycorp.service.company.model.CompanyStatus extends java.lang.Enum<com.mycorp.service.company.model.CompanyStatus> {
```

**Comment by FredDeschenes** [comment]
Yeah it's super weird for the Kotlin compiler to be marking the list as it's doing since it compiles the enum just like javac would : `public final class eu.psartini.issues.jdbi.MyEnum extends java.lang.Enum<eu.psartini.issues.jdbi.MyEnum>`, but at least there's a couple of workarounds for that (using an array or the `@JvmSuppressWildcards` annotation).

As for the return type/mapping it's fine as it's compiling it just like we'd expect (`List<MyEnum>`), anyway as you said JDBI needs an actual type to map to.

**Comment by qualidafial** [comment]
@stevenschlansker not strictly speaking. Enums are "effectively" final but individual constants can be anonymous inner classes.

@psartini Can you build Jdbi from master, and tell me whether `kotlin.collections.List<MyEnum>` works with that code? I suspect it will, because `EnumSqlArrayTypeFactory` makes it decision based on the erased type, not the full generic type signature.

**Comment by psartini** [comment]
@qualidafial not sure when I will be able to try this - don't count on this for the next 2 weeks... at least.

**Comment by qualidafial** [comment]
@psartini This is now released in 3.7.0 if you want to give it a spin

**Comment by stevenschlansker** [comment]
I added a test case in #1596 - please let us know if there's any additional problems here, otherwise this issue will be closed.  Thanks for using Jdbi!
